### PR TITLE
BigEarthNet Trainers

### PR DESCRIPTION
**Pull request by isaaccorley**
- Adds `torchgeo.trainers.BigEarthNetClassificationTask` and `torchgeo.trainers.BigEarthNetDataModule`
- Adds `torchgeo.trainers.tasks.MultiLabelClassficationTask`. This differs from `ClassificationTask` by using `torch.nn.BCEWithLogitsLoss` and modifies the metrics to handle multilabel outputs
- Adds BigEarthNet train configs
- Adds unit tests
- Added additional BigEarthNet test data to allow for a train/val/test split
- Adds estimate band min/max stats based on random samples from the dataset

TODO:
- Need to train to get benchmark result and compare to recent papers

### CODE REVIEW

File: tests/trainers/test_bigearthnet.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import pytest
+from _pytest.fixtures import SubRequest
+
+from torchgeo.trainers import BigEarthNetDataModule
+
+
+@pytest.fixture(scope="module", params=[("s1", 2), ("s2", 12), ("all", 14)])
+def bands(request: SubRequest) -> Tuple[str, int]:
+    return cast(Tuple[str, int], request.param)
```
Comment (adamjstewart) [inline tests/trainers/test_bigearthnet.py:15]: This isn't needed, you can just add these params to the `datamodule` fixture.

File: torchgeo/trainers/__init__.py [hunk 2]
```diff
@@ -11,13 +12,16 @@
...
 from .ucmerced import UCMercedClassificationTask, UCMercedDataModule
 
 __all__ = (
     # Tasks
     "ClassificationTask",
+    "MultiLabelClassificationTask",
     # Trainers
+    "BigEarthNetClassificationTask",
+    "BigEarthNetDataModule",
     "BYOLTask",
```
Comment (adamjstewart) [inline torchgeo/trainers/__init__.py:25]: Many of these are also Tasks. In my mind, there are two types of Trainers: Tasks and DataModules. Or does PyTorch-lightning have some different meaning than what I'm thinking of?

File: conf/bigearthnet.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  module:
+    loss: "bce"
+    classification_model: "resnet18"
+    learning_rate: 1e-3
+    learning_rate_schedule_patience: 6
+    in_channels: 14
+  datamodule:
+    batch_size: 128
+    num_workers: 6
+    bands: "all"
```
Comment (adamjstewart) [inline conf/bigearthnet.yaml:18]: Most of these two conf files are duplicated, should they be?

File: tests/trainers/test_tasks.py [hunk 4]
```diff
@@ -9,19 +9,31 @@
...
+    ClassificationTask,
+    MultiLabelClassificationTask,
+    So2SatDataModule,
+)
 
 from .test_utils import mocked_log
 
 
 @pytest.fixture(scope="module", params=[("rgb", 3), ("s2", 10)])
-def bands(request: SubRequest) -> Tuple[str, int]:
```
Comment (adamjstewart) [inline tests/trainers/test_tasks.py:18]: Since these bands and datamodule fixtures are specific to the task being tested, let's move them into the class definition, then their names don't need to change. For example, `bands_so2sat` would be under the `TestClassificationTask` class and be called `bands`.

File: torchgeo/trainers/bigearthnet.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+from ..datasets import BigEarthNet
+from ..datasets.utils import dataset_split
+from .tasks import MultiLabelClassificationTask
+
+# https://github.com/pytorch/pytorch/issues/60979
+# https://github.com/pytorch/pytorch/pull/61045
+DataLoader.__module__ = "torch.utils.data"
+
+
+class BigEarthNetClassificationTask(MultiLabelClassificationTask):
```
Comment (adamjstewart) [inline torchgeo/trainers/bigearthnet.py:22]: Why does this need to be a subclass? Can't the `num_classes` be an attr in `__init__`?

File: torchgeo/trainers/tasks.py [hunk 6]
```diff
@@ -257,3 +261,118 @@ def configure_optimizers(self) -> Dict[str, Any]:
                 "monitor": "val_loss",
             },
         }
+
+
+class MultiLabelClassificationTask(ClassificationTask):
+    """Abstract base class for multi label image classification LightningModules."""
```
Comment (adamjstewart) [inline torchgeo/trainers/tasks.py:267]: This is not an ABC (doesn't inherit from `abc.ABC`) and I don't think it needs to be. Why can't `num_classes` be an arg passed to `__init__`?

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Doesn't this trainer need to be added to `train.py`?

File: torchgeo/trainers/__init__.py (hunk 2 above)
Comment (isaaccorley) [reply torchgeo/trainers/__init__.py:25]: No you're right. The Tasks are LightningModules. DataModules I wouldn't consider tasks but they are grouped in the same script with the tasks. I'll break them out into tasks and datamodules.

File: conf/bigearthnet.yaml (hunk 3 above)
Comment (isaaccorley) [reply conf/bigearthnet.yaml:18]: I think the purpose is that configs in conf/task_defaults have all the required params to run, but conf/ basically inherits params from conf/task_defaults if they are not specified. So you could essentially have almost no params defined in conf/ configs and it would still run.

File: torchgeo/trainers/bigearthnet.py (hunk 5 above)
Comment (isaaccorley) [reply torchgeo/trainers/bigearthnet.py:22]: Doesn't need to be. Was just following the convention from previous trainers hardcoding in the script. Agreed that can be refactored to be called in `__init__`

File: torchgeo/trainers/tasks.py (hunk 6 above)
Comment (isaaccorley) [reply torchgeo/trainers/tasks.py:267]: It definitely can be. Will refactor to do so.

File: conf/bigearthnet.yaml (hunk 3 above)
Comment (adamjstewart) [reply conf/bigearthnet.yaml:18]: In that case why do we need two separate files at all? We should have a single file with defaults and users can provide additional files or specify on the CLI if they want to override. @calebrob6's explanation didn't really make sense to me.

Comment (calebrob6) [reply conf/bigearthnet.yaml:18]: What part didn't make sense? The `task_defaults/` are default parameters for the trainers. The others are scripts that you can use to run the trainers.

We could reorganize to have `conf/task_defaults/` and `conf/examples/`. Alternatively, you could rewrite the task defaults as dataclasses per https://omegaconf.readthedocs.io/en/2.1_branch/structured_config.html and include them in the code somewhere.

Comment (adamjstewart) [reply conf/bigearthnet.yaml:18]: > The others are scripts that you can use to run the trainers.

But couldn't you just use the files in `conf/task_defaults` for this? At least from this PR they seem to have the same structure and mostly the same values.

Comment (calebrob6) [reply conf/bigearthnet.yaml:18]: Its a convenience thing - you could just not have defaults at all and have to specify everything all the time - I usually prefer to draw the line closer to "will run by default" (these will not run by default because you need to specify the dataset directory). Take a look through all the config and how they map to arguments in `train.py` and the tasks/datamodules.

Comment (adamjstewart) [reply conf/bigearthnet.yaml:18]: I still wish `./train.py --help` told me which variables I can actually set. I usually have to dig through these conf files, and even then I don't always find the settings I need...

Comment (calebrob6) [reply conf/bigearthnet.yaml:18]: But you literally haven't used train.py lol -- currently you just don't like the _idea_ that it doesn't tell you which variables can be set. If you want to tear up the current omegaconf way of doing things, replace it with argparse, and make a way where we can still pass arguments as a yaml files then I'm totally with it.

Comment (isaaccorley) [reply conf/bigearthnet.yaml:18]: Just my 2 cents, but using config files and omegaconf or hydra is better than argparse as the number of args grows you'll end up just making bash scripts to save which args you used for a run which is effectively uglier than using config files. 

We may be able to get the best of both however by using the "add_model_specific_args" method in the tasks which we can add to the argparse in train.py https://pytorch-lightning.readthedocs.io/en/latest/common/hyperparameters.html.

Comment (adamjstewart) [reply conf/bigearthnet.yaml:18]: I definitely agree that config files are better than no config files, but if we can get both config files and a good way to document which settings can be modified, that would be even better!

Comment (isaaccorley) [reply conf/bigearthnet.yaml:18]: Yeah I think adding args to the lightning modules would be incredibly easy by adding the method above to the base tasks. This way they show up in train.py --help and we can keep track in the scripts.

Comment (isaaccorley) [reply conf/bigearthnet.yaml:18]: Going to leave as is for now. Will submit a new PR adding hparam argparse args.

File: torchgeo/trainers/tasks.py (hunk 6 above)
Comment (isaaccorley) [reply torchgeo/trainers/tasks.py:267]: There's not a great way to do this w/o adding more boilerplate everywhere. So I'll add this in a future PR.

Comment (adamjstewart) [reply torchgeo/trainers/tasks.py:267]: Can you explain this in more detail? Why can't it just be an arg to `__init__`?

File: tests/trainers/test_bigearthnet.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import pytest
+from _pytest.fixtures import SubRequest
+
+from torchgeo.trainers import BigEarthNetDataModule
+
+
+class TestBigEarthNetDataModule:
+    @pytest.fixture(
+        params=[
+            ("s1", True),
```
Comment (adamjstewart) [inline tests/trainers/test_bigearthnet.py:16]: Let's use `zip(["s1", "s2", "all"], [True, True, False])` here to cut down on the total number of combinations we test.

Comment (adamjstewart) [reply tests/trainers/test_bigearthnet.py:16]: Also add `scope="class"` so we don't have to reinstantiate the dm each time.

File: tests/trainers/test_tasks.py [hunk 8]
```diff
@@ -9,30 +9,37 @@
...
-    return dm
-
-
 class TestClassificationTask:
+    @pytest.fixture(params=[("rgb", 3), ("s2", 10)])
+    def bands(self, request: SubRequest) -> Tuple[str, int]:
+        return cast(Tuple[str, int], request.param)
+
+    @pytest.fixture(params=[True, False])
+    def datamodule(
```
Comment (adamjstewart) [inline tests/trainers/test_tasks.py:28]: I'm not sure if we actually need to parametrize the datamodule by bands or unsupervised mode. We already parametrize the datamodule during datamodule testing, here we're only trying to ensure that `ClassificationTask` works as expected. This will cut down on total testing time by a lot, especially since we're already parametrizing on loss and weight init.

File: tests/trainers/test_tasks.py [hunk 9]
```diff
@@ -110,3 +117,85 @@ def test_invalid_pretrained(self, checkpoint: str, config: Dict[str, Any]) -> No
...
         with pytest.raises(ValueError, match=error_message):
             ClassificationTask(**config)
+
+
+class TestMultiLabelClassificationTask:
+    @pytest.fixture(params=[("s1", 2), ("s2", 12), ("all", 14)])
+    def bands(self, request: SubRequest) -> Tuple[str, int]:
+        return cast(Tuple[str, int], request.param)
+
+    @pytest.fixture(params=[True, False])
```
Comment (adamjstewart) [inline tests/trainers/test_tasks.py:127]: Same here, let's skip all the datamodule parametrization

File: torchgeo/trainers/__init__.py [hunk 10]
```diff
@@ -11,29 +12,33 @@
 from .resisc45 import RESISC45ClassificationTask, RESISC45DataModule
 from .sen12ms import SEN12MSDataModule, SEN12MSSegmentationTask
 from .so2sat import So2SatClassificationTask, So2SatDataModule
-from .tasks import ClassificationTask
+from .tasks import ClassificationTask, MultiLabelClassificationTask
 from .ucmerced import UCMercedClassificationTask, UCMercedDataModule
 
 __all__ = (
-    # Tasks
+    # Base Classes
```
Comment (adamjstewart) [inline torchgeo/trainers/__init__.py:19]: Get rid of "Base Classes" and just call them all "Tasks"

**Review by adamjstewart — CHANGES_REQUESTED** [review]
I still haven't heard a good reason why we can't have a single `ClassificationTask` and `MultiLabelClassificationTask` without all of these subclasses.

File: tests/trainers/test_tasks.py (hunk 8 above)
Comment (isaaccorley) [reply tests/trainers/test_tasks.py:28]: We need to test by bands because the min/max tensors we use for normalizing change based on the bands in the datamodule. If we create a base Classification DataModule we can test `unsupervised_mode` on that and not have to repeat for every datamodule.

Comment (isaaccorley) [reply tests/trainers/test_tasks.py:28]: Just a heads up, I used `tests/trainers/test_so2sat.py` as a template for this so that will need to refactored later as well.

**Comment by isaaccorley** [comment]
I refactored test_tasks to use a dummy dataset and datamodule so we can test for varying number of channels but avoid having to repeat tests for bigearthnet and so2sat datamodule specific args. Let me know if this works.

**Comment by isaaccorley** [comment]
You are right thought that there is not a good case for making a separate task for each classification dataset. We can definitely remove them if that's the route you want to take.

File: torchgeo/trainers/bigearthnet.py (hunk 5 above)
Comment (adamjstewart) [inline torchgeo/trainers/bigearthnet.py:22]: I think getting rid of this class is the last remaining thing to do before this can be merged.

Comment (adamjstewart) [reply torchgeo/trainers/bigearthnet.py:22]: Personally I would refactor `ClassificationTask` first in a separate PR to keep this PR from getting any bigger.

Comment (isaaccorley) [reply torchgeo/trainers/bigearthnet.py:22]: We should merge this because I still need to refactor the num classes as an input arg before I can remove it. If I do that then I will need to remove/refactor all other classification tasks as well.

**Review by adamjstewart — APPROVED** [review]
